src/crimson/osd/pg_backend.cc

```diff
@@ -256,6 +256,49 @@ bool PGBackend::maybe_create_new_object(
   return true;
 }
 
+seastar::future<> PGBackend::write(
+    ObjectState& os,
+    OSDOp& osd_op,
+    ceph::os::Transaction& txn)
+{
+  ceph_osd_op& op = osd_op.op;
+  if (auto seq = os.oi.truncate_seq;
+      seq != 0 && op.extent.truncate_seq < seq) {
+    // old write, arrived after trimtrunc
+    if (op.extent.offset + op.extent.length > os.oi.size) {
+      // no-op
+      if (op.extent.offset > os.oi.size) {
+	op.extent.length = 0;
+	osd_op.indata.clear();
+      } else {
+	// truncate
+	auto len = os.oi.size - op.extent.offset;
+	osd_op.indata.splice(len, op.extent.length);
+	op.extent.length = len;
+      }
+    }
+  } else if (op.extent.truncate_seq > seq) {
+    // write arrives before trimtrunc
+    if (os.exists && !os.oi.is_whiteout()) {
+      txn.truncate(coll->cid, ghobject_t{os.oi.soid}, op.extent.truncate_size);
+    }
+    os.oi.truncate_seq = op.extent.truncate_seq;
+    os.oi.truncate_size = op.extent.truncate_size;
+  }
+  maybe_create_new_object(os, txn);
+  if (op.extent.length == 0) {
+    if (op.extent.offset > os.oi.size) {
+      txn.truncate(coll->cid, ghobject_t{os.oi.soid}, op.extent.offset);
+    } else {
+      // XXX: what is trans.nop() for?
```

Comment:
I think there's a check somewhere for an empty transaction, it probably ensures that the oi.version gets bumped and that it goes out as a write.

Reply:
```
prepare_transaction()
    do_osd_ops()
        // update transaction
    if (txn.empty()):
        return
    finish_ctx()
        if (new_obs.exists):
            new_os.oi.version = os.at_version
            new_os.oi.prior_versoin = os.oi.version
            attrs = {OI_ATTR:encode(new_os.oi)}
            if (soid.is_head()):
                attrs[SS_ATTR] = new_os.snapset
            txn.setattrs(attrs)
```

yes. `prepare_transaction()` updates oi if we append a nop to transaction. otherwise it will just return. but why shall we update `oi.version` for a no-op? i think we can skip this step if a new object is not created by `maybe_create_new_object()`.

Reply:
It might be important for dup detection that even a logical noop be recorded as a mutation.

Reply:
fixed and updated.